private/buf/cmd/buf/command/decode/decode.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package decode
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufreflect"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	errorFormatFlagName = "error-format"
+	sourceFlagName      = "source"
+	typeFlagName        = "type"
+	outputFlagName      = "output"
+	outputFlagShortName = "o"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name,
+		Short: "Decode binary serialized message with a source reference.",
```

Comment:
@amckinney and I are thinking about the UI here and wondering if there's one command we can have that can take an arbitrary input of either json/bin format and produce an output in another specified format. So i.e. `buf encode --input - --output message.json`. We can't figure out a good UX for this yet though. Let's get this into `buf beta` and then iterate on this. Bug @amckinney and I about this

Reply:
I am not sure about the use case for this as I think the user know what they want to do (encode/decode) in most of the time? but yes let's discuss this 🙌